test/files/run/t10513.scala

```diff
@@ -0,0 +1,34 @@
+//package scala.concurrent
+import scala.concurrent._
+
+import scala.util.{Random, Try}
+import ExecutionContext.Implicits.global
+
+/** This test uses recursive calls to Future.flatMap to create arrays whose
+ *  combined size is slightly greater than the JVM heap size. A previous
+ *  implementation of Future.flatMap would retain references to each array,
+ *  resulting in a speedy OutOfMemoryError. Now, each array should be freed soon
+ *  after it is created and the test should complete without problems.
+ */
+object Test {
+ val arrSz = 50 * 10000
+  val numFutures = 10000
+
+  val rng = new Random()
+
+
+  def main(args: Array[String]) {
+    val longStandingPromise = Promise[Nothing]
+
+    val futures = List.tabulate(numFutures){ i =>
+      val arr = Array.tabulate(arrSz)(identity)
+      val idx = rng.nextInt(arrSz)
+      val f1 = Future{ arr }
+      val f2 = Future.firstCompletedOf(List(longStandingPromise.future, f1))
+      f2.map( arr => arr(idx))
+    }
+    val fSeq = Future.sequence(futures)
+    val finalF = fSeq.map(_.sum)
+    val res = Await.result(finalF, duration.Duration.Inf)
```

Comment:
This is a bit of a recipe for a hanging buildserver, we need a sensible timeout.

Reply:
checked on my 4 core 16GB laptop, original failing test fails after 20 seconds, same test with the fix applied completed after 74 seconds.
so few conclusions here:
1. we can definitely decrease the number of iterations required for this test to make its point.
2. we can calculate this magic number of iteration based on the amount of memory allocated for this test, @viktorklang  can you please point me to the section in SBT that determines that?

Reply:
just took a closer look at partest's output:
> Java options are:    -Xmx1024M -Xms64M

so I guess I can set the number of iteration such that total memory consumption is ~2gb, seems 4000 iterations are enough.

Reply:
@eyalfa 10mins would be ok too :)

Reply:
partest can use a `.javaopts` file, there is a `./test/files/run/bridges.javaopts`. Also wondered if this can be written as a junit test; `test/junit/scala/collection/mutable/MutableListTest.scala` uses `assertNotReachable`, but maybe that's not practical here. There were previous tests around memory leaks, but I don't remember which offhand.